README.md
# Slack messenger


### Описание

Celery-сервис для отправки сообщений в Slack


## используемые переменные окружения

```
VERSION=1.0.0 // версия сервиса
RABBITMQ_DEFAULT_USER=admin // учетная запись для доступа к сервису rabbitmq 
RABBITMQ_DEFAULT_PASS=11111 // пароль к учетной записи для доступа к сервису rabbitmq
RABBITMQ_DEFAULT_PORT=5672 // порт по которому доступен сервис rabbitmq
RABBITMQ_DEFAULT_HOST=172.16.0.1 // хост по которому доступен сервис rabbitmq
QUEUE=slack_messenger_service.tasks // Название очереди
REDIS_HOST=172.17.0.1  // хост по которому доступен сервис redis
REDIS_PORT=6379 // порт по которому доступен сервис redis
CONCURRENCY=2 // Количество параллельных рабочих процессов
SLACK_BOT_TOKEN // Токен slack бота
```

## пример отправки задачи на отправку сообщения
```py
celery.send_task(
    "slack_messenger_service.send_message",
    queue='slack_messenger_service.tasks',
    args=(msg_channel, msg_text)
)
```

## пример запроса на отправку сообщения
```py
curl -X POST http://url:8000/api/v1/tasks/ \
-H 'Content-Type: application/json' \
-d '{
"task_name": "slack_messenger_service.send_message",
"task_args": {
"text": "test message",
"channel": "11111" // id канала 
}}'

пример ответа сервиса:

{"status":"ok","data":{"task_id":"3319bfbc-7af7-402d-80a6-e404db2d3ff4"},"debug":null,"error":null}
```


CHANGELOG.md

### 1.0.0
- реализация таска отправки сообщения в канал


tests/test_task.py
from celery import Celery
from celery.result import allow_join_result

from app.config import service_config

test_celery = Celery(service_config.service_name,
                     broker=service_config.broker,
                     backend=service_config.backend,
                     result_serializer='pickle',
                     accept_content=['pickle', 'json']
                     )


def test_run_task():
    msg_channel = 'C02EYS7TYEP'
    msg_text = '''Битые пути в company/770-plug
/goAnsOperator/Ans_NoWayToPlug  ========> file: company_770-plug/src/SPEC/Services/_goAnsServices.sc
-----------
WARNING:
/comMod_BullsAndCowsGame/BullsAndCows  ========> file: company_770-plug_dep/common/src/COMMON/_goAnsGames.sc
/comMod_KtoZhivetGame/StartGame  ========> file: company_770-plug_dep/common/src/COMMON/_goAnsGames.sc'''

    async_result = test_celery.send_task(
        f"{service_config.service_name}.send_message",
        queue=service_config.queue,
        args=(msg_channel, msg_text)
    )
    print(async_result.result)
    with allow_join_result():
        result = async_result.get()

    print(result)
    assert isinstance(result, dict)


app/tasks.py
from typing import Dict

from slack_sdk import WebClient

from base_response import BaseResponse, Status
from celery_app import celery
from config import service_config as config
from logger import logger

slack_client = WebClient(token=config.slack_bot_token)


@celery.task(name=f"{config.service_name}.send_message",
             queue=f"{config.queue}")
def send_message(channel: str, text: str) -> Dict[str, str]:
    """таск для отправки сообщений в slack

    Args:
        channel (str): ИД канала
        text (str): Текст сообщения
    """
    response = BaseResponse(
        status=Status.STATUS_OK,
        data="",
        debug="",
        error=""
    )
    logger.info(f"Start task with params channel={channel}, text={text}")

    try:
        slack_client.chat_postMessage(channel=channel, text=text)
    except Exception as e:
        response.status = Status.STATUS_ERROR
        response.error = type(e).__name__
        response.debug = str(e)

    logger.info("End execute task")

    logger.info(response.dict())

    return response.dict()


app/logger.py
import logging
from logging.handlers import TimedRotatingFileHandler
from pathlib import Path

from config import service_config as config, Env


def init_logger(service_name: str, log_level: str) -> logging.Logger:
    logger = logging.getLogger(service_name)
    logger.setLevel(getattr(logging, log_level))
    formatter = logging.Formatter(
        f"%(asctime)s.%(msecs)03dZ [{service_name}] %(levelname)s: %(message)s",
        "%Y-%m-%dT%H:%M:%S",
    )

    console_handler = logging.StreamHandler()
    console_handler.setLevel(getattr(logging, log_level))
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)

    if config.env != Env.LOCAL:
        log_path = Path(f"/var/log/{service_name}/{service_name}.log")
        rotating_handler = TimedRotatingFileHandler(
            log_path, when="midnight", encoding="utf-8"
        )
        rotating_handler.setLevel(getattr(logging, log_level))
        rotating_handler.setFormatter(formatter)
        logger.addHandler(rotating_handler)

    return logger


logger = init_logger(config.service_name, config.log_level)


app/base_response.py
from enum import Enum
from typing import Optional, Any, Union, Dict

from pydantic import BaseModel


class Status(str, Enum):
    STATUS_OK = "ok"
    STATUS_ERROR = "error"

    def __str__(self):
        """
        метод для получения строкового литерала статуса
        :return:
        """
        return self.value


class BaseResponse(BaseModel):
    """
    Базовая структура ответа сервиса.
    Все структуры ответов должны наследоваться от этого класса и переопределять параметр data.
    status - всегда либо ok, либо error
    data - полезная нагрузка, переопределяется структурой самой ручки
    error - слаг ошибки, крайне рекомендуется не отдавать сырой ответ
    debug - структура (или строка) для более подробного описания ошибки
    """

    status: Status
    data: Optional[str]
    debug: Optional[str]
    error: Optional[str]

    class Config:
        use_enum_values = True


app/config.py
import os
from enum import Enum
from pathlib import Path

from pydantic import BaseModel


class Env(str, Enum):
    LOCAL = "local"
    DEV = "dev"
    PREPROD = "preprod"
    PROD = "prod"


class Config(BaseModel):
    service_name: str
    broker: str
    backend: str
    queue: str
    version: str
    log_dir_path: Path
    log_level: str
    env: Env
    slack_bot_token: str


def init_config() -> Config:
    """
    Метод для инициализации параметров сервиса
    :return: Config, экземпляр класса с параметрами
    """
    # название сервиса, нужно указать свое при использовании шаблона
    service_name: str = os.getenv('SERVICE_NAME', 'slack_messenger_service')
    # параметры подключения к rabbitmq
    rabbit_user: str = os.getenv('RABBITMQ_DEFAULT_USER')
    rabbit_pass: str = os.getenv('RABBITMQ_DEFAULT_PASS')
    rabbit_host: str = os.getenv('RABBITMQ_DEFAULT_HOST', 'rabbitmq')
    rabbit_port: int = int(os.getenv('RABBITMQ_DEFAULT_PORT', 5672))
    queue = os.getenv("QUEUE", "slack_messenger_service.tasks")

    # параметры подключения к redis
    redis_host: str = os.getenv('REDIS_HOST', 'redis')
    redis_port: int = int(os.getenv('REDIS_PORT', 6379))

    broker: str = f"pyamqp://{rabbit_user}:{rabbit_pass}@{rabbit_host}:{rabbit_port}"
    backend: str = f"redis://{redis_host}:{redis_port}"
    log_level = os.getenv("LOG_LEVEL", "DEBUG")
    slack_bot_token = os.getenv("SLACK_BOT_TOKEN")

    return Config(
        service_name=service_name,
        broker=broker,
        backend=backend,
        version=os.getenv("VERSION", "0.0.0"),
        log_dir_path=os.getenv('LOG_DIR', f"/var/log/{service_name}/"),
        log_level=log_level,
        queue=queue,
        slack_bot_token=slack_bot_token,
        env=Env(os.getenv("ENV", Env.LOCAL))
    )


service_config = init_config()
